src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -1606,17 +1606,27 @@ protected URL findResource(String name) {
      * </ol>
      * <p>Note that once a class loader is registered as parallel capable, there
      * is no way to change it back.</p>
+     * <p>
+     * In cases where {@code ClassLoader.registerAsParallelCapable} is called from a context where
+     * there is no caller frame on the stack (e.g. when called directly
+     * from a JNI attached thread), {@code IllegalCallerException} is thrown.
+     * </p>
      *
      * @return  {@code true} if the caller is successfully registered as
      *          parallel capable and {@code false} if otherwise.
+     * @throws IllegalCallerException if there is no caller frame on the stack.
      *
      * @see #isRegisteredAsParallelCapable()
      *
      * @since   1.7
      */
     @CallerSensitive
     protected static boolean registerAsParallelCapable() {
-        return registerAsParallelCapable(Reflection.getCallerClass());
+        final Class<?> caller = Reflection.getCallerClass();
+        if (caller == null) {
```

Comment:
```suggestion
        if (caller == null || !ClassLoader.class.isAssignableFrom(caller)) {
             throw new IllegalCallerException(caller + " not a subclass of ClassLoader");
        }     
```

What we suggested is to throw IllegalCallerException if the caller is not a class loader and that will include null caller case.

Reply:
This also needs a test case for non-class loader caller.

I looked at the existing tests testing `registerAsParallelCapable` but I can't find a good one to add the new test case.   The closest one could be `test/jdk/java/lang/ClassLoader/IsParallelCapable.java`.   I'm okay to include a new test case in `IsParallelCapable.java` to verify `IllegalCallerException` if called by a non-class loader.